src/main/java/com/zendesk/maxwell/bootstrap/SynchronousBootstrapper.java

```diff
@@ -0,0 +1,212 @@
+package com.zendesk.maxwell.bootstrap;
+
+import com.google.code.or.OpenReplicator;
+import com.zendesk.maxwell.BinlogPosition;
+import com.zendesk.maxwell.MaxwellContext;
+import com.zendesk.maxwell.MaxwellReplicator;
+import com.zendesk.maxwell.RowMap;
+import com.zendesk.maxwell.producer.AbstractProducer;
+import com.zendesk.maxwell.schema.Database;
+import com.zendesk.maxwell.schema.Schema;
+import com.zendesk.maxwell.schema.Table;
+import com.zendesk.maxwell.schema.columndef.ColumnDef;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.sql.*;
+import java.util.Iterator;
+
+public class SynchronousBootstrapper extends AbstractBootstrapper {
+
+	static final Logger LOGGER = LoggerFactory.getLogger(MaxwellReplicator.class);
+	private static final long INSERTED_ROWS_UPDATE_PERIOD_MILLIS = 250;
+
+	private long lastInsertedRowsUpdateTimeMillis = 0;
+
+	public SynchronousBootstrapper(MaxwellContext context) { super(context); }
+
+	@Override
+	public boolean isStartBootstrapRow(RowMap row) {
+		return isBootstrapRow(row) &&
+			row.getData("started_at") == null &&
+			row.getData("completed_at") == null &&
+			( long ) row.getData("is_complete") == 0;
+	}
+
+	@Override
+	public boolean isCompleteBootstrapRow(RowMap row) {
+		return isBootstrapRow(row) &&
+			row.getData("started_at") != null &&
+			row.getData("completed_at") != null &&
+			( long ) row.getData("is_complete") == 1;
+	}
+
+	@Override
+	public boolean isBootstrapRow(RowMap row) {
+		return row.getDatabase().equals("maxwell") &&
+			row.getTable().equals("bootstrap");
+	}
+
+	@Override
+	public boolean shouldSkip(RowMap row) {
+		// the synchronous bootstrapper blocks other incoming messages
+		// to the replication stream so there's nothing to skip
+		return false;
+	}
+
+	@Override
+	public void startBootstrap(RowMap startBootstrapRow, AbstractProducer producer, MaxwellReplicator replicator) throws Exception {
+		String databaseName = ( String ) startBootstrapRow.getData("database_name");
+		String tableName = ( String ) startBootstrapRow.getData("table_name");
+		LOGGER.debug(String.format("bootstrapping request for %s.%s", databaseName, tableName));
+		Schema schema = replicator.getSchema();
+		Database database = findDatabase(schema, databaseName);
+		Table table = findTable(tableName, database);
+		OpenReplicator openReplicator = replicator.getOpenReplicator();
+		BinlogPosition position = new BinlogPosition(openReplicator.getBinlogPosition(), openReplicator.getBinlogFileName());
+		producer.push(startBootstrapRow);
+		producer.push(bootstrapStartRowMap(table, position));
+		LOGGER.info(String.format("bootstrapping started for %s.%s, binlog position is %s", databaseName, tableName, position.toString()));
+		try ( Connection connection = getConnection() ) {
+			setBootstrapRowToStarted(startBootstrapRow, connection);
+			ResultSet resultSet = getAllRows(databaseName, tableName, schema, connection);
```

Comment:
I'm pretty sure I understand, but I just wanted to check:

This does a *single* "select *" command, right? And then JDBC (may) chunk it into bootstrap_fetch_size chunks internally?

It doesn't do multiple "select *" commands? Because that wouldn't work, consistency-wise. You want the bootstrap to be a single consistent snapshot.